lib/pure/hashes.nim

```diff
@@ -70,6 +70,46 @@ proc `!$`*(h: Hash): Hash {.inline.} =
   res = res + res shl 15
   result = cast[Hash](res)
 
+proc hiXorLoFallback(A, B: uint64): uint64 {.inline.} =
+  let # Fall back for weaker platforms/compilers, e.g. Nim VM, etc.
+    ha: uint64 = A shr 32
+    hb: uint64 = B shr 32
+    la: uint64 = A and 0xFFFFFFFF'u64
+    lb: uint64 = B and 0xFFFFFFFF'u64
+    rh: uint64 = ha * hb
+    rm0: uint64 = ha * lb
+    rm1: uint64 = hb * la
+    rl: uint64 = la * lb
+    t: uint64 = rl + (rm0 shl 32)
+  var c = if t < rl: 1'u64 else: 0'u64
+  let lo = t + (rm1 shl 32)
+  c += (if lo < t: 1'u64 else: 0'u64)
+  let hi = rh + (rm0 shr 32) + (rm1 shr 32) + c
+  return hi xor lo
+
+proc hiXorLo(A, B: uint64): uint64 {.inline.} =
+  # Xor of high & low 8B of full 16B product
+  when nimvm:
+    result = hiXorLoFallback(A, B) # `result =` is necessary here.
+  else:
+    when Hash.sizeof < 8:
+      result = hiXorLoFallback(A, B)
+    elif defined(gcc) or defined(llvm_gcc) or defined(clang):
+      {.emit: """__uint128_t r = A; r *= B; return (r >> 64) ^ r;""".}
+    elif defined(windows) and not defined(tcc):
+      {.emit: """A = _umul128(A, B, &B); return A ^ B;""".}
+    else:
+      result = hiXorLoFallback(A, B)
+
+proc hashWangYi1*(x: int64|uint64|Hash): Hash {.inline.} =
+  ## Wang Yi's hash_v1 for 8B int.  https://github.com/rurban/smhasher has more
+  ## details.  This passed all scrambling tests in Spring 2019 and is simple.
+  ## NOTE: It's ok to define ``proc(x: int16): Hash = hashWangYi1(Hash(x))``.
+  const P0 = 0xa0761d6478bd642f'u64
+  const P1 = 0xe7037ed1a0b428db'u64
+  const P5x8 = 0xeb44accab455d165'u64 xor 8'u64
+  Hash(hiXorLo(hiXorLo(P0, uint64(x) xor P1), P5x8))
```

Comment:
This should be a ``cast[Hash]``, I think.

Reply:
Ok.  I'll try that.  Yeah I was getting i386/32-bit errors in the CI but there was no line number in the range exception.  So, I was kind of guessing my way around.  Thanks for the pointer!  Pushed an attempt to fix.  We'll see how it goes.

Reply:
Well, that `cast` instead of convert seems to have fixed the Linux i386.

Any ideas why `tests/closure/tclosure.nim` would be failing on the CI?  It works for me locally on an Linux_amd64 system, but seems to stop between the end of `block doNotation` and `block tforum`.  My guess would be `block fib50` since that engages the new `hash()`, but I don't get an assertion error from the CI.

Reply:
Well, though that `tclosure.nim` failure had repeated quite a few times it now appears to have worked and Linux i386 passes all tests.

The testament `nim pcat collections` seems to be infinite looping holding up other people's CI's, and I am unsure why, and I do not see how to kill my CI jobs in the UI.  So you should kill them which I think you (or someone) also did yesterday.

Also, before merging this we still need to decide on a `defined(js)` branch.  It seems the arithmetic JS does creates zero output hashes with `hiXorLoFallback64`.  Over at https://github.com/c-blake/adix in `althash.nim`, I have a `hiXorLoFallback32` in 32-bit arithmetic that gives bit identical hashes with the C backend, but different hashes with the JS backend (not sure why, but may be fixable) and also uses 16 multiplications and so is probably kinda slow.

Reply:
Never mind Re killing jobs.  I guess they timeout at 90 minutes.  We do need to figure out why they infinite loop, though, and also decide about `defined(js)`.

Reply:
I'm investigating.

Reply:
Ok.  Thanks!

Reply:
I found a decent answer for the JS part that gives the same function (in low 32-bits anyway, which is what sets & tables care about).  Might not be hard to extend that to 53 bits if people complain.  Getting all the way to 64 bits will require changing the backend to use `BigInt` rather than `Number` for `int64`/`uint64`, but I doubt anyone is using even 4 gigaEntry int-keyed tables on the JS backend.  That'd be at least 64 GiB.